docs/DMS/PRIMARY-DATA-STORAGE.md

```diff
@@ -25,3 +25,268 @@ Braindump:
+
+One question is how performant does this need to be? If/when it becomes too slow, moving to a search engine would be the recommendation because if you want true performance you would use a separate read only store.
+
+### Security
+
+The next thing that you need to be able to support is security. Following the ODS/API, we'll want Namespace-based for sure, Education Organization-based probably, and possibly Grade Level-based as well. In some ways this may be similar to query support, as we'll need to know the relevant fields to secure on for each resource.
+
+### Streaming changes
+
+We'll need the ability to stream out changes. This needs to be something that can be easily integrated with Kafka and potentially other streaming platforms, with a well-defined message format. Change query support is related to this, but we won't prejudge whether it will be based off of streaming.
+
+### Simplicity
+
+As maintainability is a primary goal, we want to keep Tanager as simple as possible. Meadowlark demonstrated that we can simplify over the ODS/API by avoiding code-per-resource and instead use a generalized design applicable to any resource. While it may not be possible with Tanager to do this completely, we would like to do it whenever possible.
+
+### Performance
+
+We need to make sure that we design to meet performance goals, and test early and often to ensure that we are able to hit those goals.
+
+## Potential Solutions
+
+### Option A - Similar to Meadowlark
+
+```mermaid
+erDiagram
+    References }o--|| Documents : ""
+    References {
+        bigint id PK "Auto-increment, otherwise unused"
+        Guid parent_ref_id FK "Parent document with reference"
+        Guid referenced_ref_id FK "Document being referenced"
+    }
+    Documents {
+        Guid referential_id PK "Hash of identity"
+        Guid document_uuid "API resource id, indexed"
+        string project_name "Example: Ed-Fi (for DS)"
+        string resource_name "Example: Student"
+        string resource_version "Example: 5.0.0"
+        JSON edfi_doc "The document"
+    }
+```
+
+This is very much like the Meadowlark implementation of the PostgreSQL backend, except we want to take advantage of foreign key constraints for reference validation.
+
+We would have a `Documents` table that will hold all of the documents for all of the entities. This will be a partitioned table so that we don't have to be concerned about how big this table grows. The documents table will have a `referential_id` as both the primary key and a hash partition key. It will be a UUIDv5 (see [RFC 4122](https://datatracker.ietf.org/doc/html/rfc4122#section-4.3)) with an Ed-Fi namespace ID and the resource name + the extracted document identity as the "name". This concept of a deterministic hash UUID allows Tanager to determine both document identities and document references independent of data in the DB. This ID will not be provided to API clients.
```

Comment:
Partitioning will be incredibly important here and probably worth describing in a little more detail. For example, how many partitions by default? The default data standard has something on the order of 70 +/- 10 resources in it, if I remember correctly. Some of those won't be written to. Can we do something like 2^6 (64) partitions by default? (power of 2 may be useful when thinking about partitioning based on guid, which uses hexadecimal).